src/node_large_page.cc

```diff
@@ -0,0 +1,233 @@
+      ret = fscanf(f, "%lx-%lx %4s %lx %5s %ld %s\n",
+                   &start, &end, perm, &offset, dev, &inode, name);
+      fclose(f);
+
+      if (ret == 7 &&
+          perm[0] == 'r' && perm[1] == '-' && perm[2] == 'x') {
+        // Checking if the region is from node binary and executable
+        start = (unsigned int64_t) &__nodetext;
+        char *from = reinterpret_cast<char *>PAGE_ALIGN_UP(start, hugePageSize);
+        char *to = reinterpret_cast<char *>PAGE_ALIGN_DOWN(end, hugePageSize);
+
+        if (from < to) {
+          size_t size = (intptr_t)to - (intptr_t)from;
+          nregion.found_text_region = true;
+          nregion.from = from;
+          nregion.to = to;
+          nregion.offset = offset;
+          nregion.totalHugePages = size/hugePageSize;
+          return nregion;
+        }
+      }
+
+      return nregion;
+    }
+
+    static bool isTransparentHugePagesEnabled() {
+       std::ifstream ifs;
+
+       ifs.open("/sys/kernel/mm/transparent_hugepage/enabled");
+       if (!ifs) {
+         fprintf(stderr, "WARNING: Couldn't check hugepages support\n");
+         return false;
+       }
+
+       std::string always, madvise, never;
+       if (ifs.is_open()) {
+          while (ifs >> always >> madvise >> never) {}
+       }
+
+       int ret_status = false;
+
+       if (always.compare("[always]") == 0)
+         ret_status = true;
+       else if (madvise.compare("[madvise]") == 0)
+         ret_status = true;
+       else if (never.compare("[never]") == 0)
+         ret_status = false;
+
+       ifs.close();
+       return ret_status;
+    }
+
+    //  Moving the text region to large pages. We need to be very careful.
+    //  a) This function itself should not be moved.
+    //     We use a gcc option to put it outside the ".text" section
+    //  b) This function should not call any function(s) that might be moved.
+    //    1. We map a new area and copy the original code there
+    //    2. We mmap using HUGE_TLB
+    //    3. If successful we copy the code there and unmap the original region.
+    int
+      __attribute__((__section__(".eh_frame")))
+      __attribute__((__aligned__(2 * 1024 * 1024)))
+      __attribute__((__noinline__))
+      __attribute__((__optimize__("2")))
+      move_text_region_to_large_pages(struct TextRegion r) {
+        void *nmem = nullptr, *tmem = nullptr;
+        int ret = 0;
+
+        size_t size = (intptr_t)r.to - (intptr_t)r.from;
+        void *start = r.from;
+
+        // Allocate temporary region preparing for copy
+        nmem = mmap(nullptr, size,
+                    PROT_READ | PROT_WRITE, MAP_PRIVATE | MAP_ANONYMOUS, -1, 0);
+        if (nmem == MAP_FAILED) {
+          printSystemError(errno);
+          return -1;
+        }
+
+        memcpy(nmem, r.from, size);
+
+        tmem = mmap(start, size,
+                    PROT_READ | PROT_WRITE | PROT_EXEC,
+                    MAP_PRIVATE | MAP_ANONYMOUS | MAP_FIXED, -1 , 0);
```

Comment:
what is the rationale behind predefining the memory attributes? ideally we want to inherit the attribute from the old small pages?

Reply:
We do want to be able to write to this so PROT_WRITE is needed. We are using MAP_FIXED to place the mapping at exactly that address.

Reply:
Updated the code comments
We already know the original page is r-xp (PROT_READ, PROT_EXEC, MAP_PRIVATE)
We want PROT_WRITE because we are writing into it.
We want it at the fixed address and we use MAP_FIXED.
We are using MAP_ANONYMOUS so it's contents are initialized to zero and it is also not backed by any file.